File: examples/postgres-lifecycle/tests/diagnostics.py
"""External watchdog for executable exit tests; no PostgreSQL is contacted."""

import os
from pathlib import Path
import signal
import sys

sys.path.insert(0, str(Path(__file__).resolve().parents[3] / "scripts"))
from scheduling_process import run_process


def require(condition, message):
    if not condition:
        raise AssertionError(message)


def observe(command, *, env=None, timeout=10):
    result = run_process(command, env=env, timeout=timeout,
                         output_limit=65536, reap_allowance=2)
    # Assert metadata only: raw failing output might contain credentials.
    require(result.reaped and result.output_eof, result.metadata())
    require(not result.overflow and not result.errors, result.metadata())
    return result


def configuration(binary):
    for value in (None, b"postgres://test:credential-marker@localhost/db\xff",
                  b"postgres://test:credential-marker@127.0.0.1:invalid-port/db"):
        env = os.environb.copy()
        env.pop(b"DATABASE_URL", None)
        if value is not None:
            env[b"DATABASE_URL"] = value
        result = observe([binary], env=env)
        require(not result.watchdog and result.status == 1, result.metadata())
        require(result.stderr == b"Error: process failed\n", "unexpected exit diagnostic")
        require(b"credential-marker" not in result.stdout + result.stderr, "credential escaped")


def missing_configuration(binary):
    env = os.environb.copy()
    env.pop(b"DATABASE_URL", None)
    result = observe([binary], env=env)
    require(not result.watchdog and result.status == 1, result.metadata())
    require(not result.stdout, "unexpected standard output")
    require(result.stderr == b"Error: process failed\n", "unexpected exit diagnostic")


def watchdog():
    result = observe([sys.executable, "-c", "import time; time.sleep(60)"], timeout=0.2)
    require(result.watchdog and result.status == -signal.SIGKILL, result.metadata())
    require(result.elapsed_seconds < 5, result.metadata())


if __name__ == "__main__":
    if sys.argv[1] == "configuration":
        configuration(sys.argv[2])
    elif sys.argv[1] == "watchdog":
        watchdog()
    elif sys.argv[1] == "missing-configuration":
        missing_configuration(sys.argv[2])
    else:
        raise SystemExit("unknown diagnostic control")

File: examples/postgres-lifecycle/tests/diagnostics.rs
use std::{path::Path, process::Command};

fn run_control(name: &str) {
    // The Python parent owns a wall-clock watchdog independently of the example
    // runtime, with bounded capture and explicit kill/reap on incomplete exit.
    let status = Command::new("python3")
        .arg(Path::new(env!("CARGO_MANIFEST_DIR")).join("tests/diagnostics.py"))
        .arg(name)
        .arg(env!("CARGO_BIN_EXE_postgres_lifecycle"))
        .status()
        .expect("external diagnostic watchdog must execute");
    assert!(status.success(), "diagnostic control failed: {name}");
}

#[test]
fn configuration_failures_exit_without_printing_credentials() {
    run_control("configuration");
}

#[test]
fn diagnostic_watchdog_kills_and_reaps_a_hung_child() {
    run_control("watchdog");
}
